man/xpak.5
.TH XPAK 5 "Oct 2022" "Portage VERSION" "Portage"
.SH NAME
xpak \- The XPAK Data Format used with Portage binary packages
.SH DESCRIPTION
Gentoo binary packages built in the \fIxpak\fR format have an xpak attached to
it which contains build time information like the USE flags it was built with,
the ebuild it was built from, the environmental variables, CFLAGS, CXXFLAGS, etc.
Note that the \fIgpkg\fR format is different and described in GLEP 78.
.SH NOTES
.SS Data Types
The following conventions cover all occurrences in this documentation
.IP Integer
All offsets/lengths are big endian unsigned 32bit integers
.IP String
All strings are ASCII encoded, and not NUL terminated (quotes are for
illustration only)
.IP Values
The actual values of the individual xpak entries are stored as Strings
.P
.SS Vertical Bars
The vertical bars '|' are not part of the file format; they are merely used to
illustrate how the offset values apply to the data.
.SH SYNOPSIS
.IP "binpkg (tbz2)"
      |<-xpak_offset->|
 <tar>|<    xpak     >|<xpak_offset>"STOP"
.IP xpak
 "XPAKPACK"<index_len><data_len><index><data>"XPAKSTOP"
.IP index
 |<-------------index_len------------->|
 |<index1><index2><index3><index4><...>|
.IP indexN
           |<-name_len->|
 <name_len>|<   name   >|<data_offset><data_len>
.IP data
 |<--------------data_len------------->|
 |<-dataN_offset->|<-dataN_len->|
 |<     data     >|<  data_N   >|<data>|
.SH DETAILS
.SS xpak

If you look at a Gentoo binary package (binpkg) with a hex-editor you'll
notice that after the tarball of files, you find a binary blob - the
\fIxpak\fR, an offset which holds the bytes from the start of the
\fIxpak\fR to the end of the file - \fIxpak_offset\fR and finally the
String \fI"STOP"\fR.

      |<xpak_offset>|
 <tar>|<---xpak---->|<xpak_offset>"STOP"

Here you see the \fItar\fR archive, the attached \fIxpak\fR blob, the
\fIxpak_offset\fR and the string \fI"STOP"\fR at the end.  This metadata
is not considered "part" of the \fIxpak\fR, but rather part of the binpkg.

If we read the offset value and count \fIoffset\fR bytes backwards from
the start of \fIxpak_offset\fR, we have found the start of the \fIxpak\fR
block which starts with the String \fI"XPAKPACK"\fR.

This xpak block consists of the string \fI"XPAKPACK"\fR, the length of the
\fIindex\fR block (\fIindex_len\fR), the length of the \fIdata\fR block
(\fIdata_len\fR), an \fIindex_len\fR bytes long binary blob with the
\fIindex\fR, a \fIdata_len\fR bytes long binary blob with the \fIdata\fR,
and the string \fI"XPAKSTOP"\fR at the end:

                                |<index_len>|<data_len>|
 "XPAKPACK"<index_len><data_len>|<--index-->|<--data-->|"XPAKSTOP"

To actually get the \fIindex\fR and the \fIdata\fR, we cut out \fIindex_len\fR
bytes after the end of \fIdata_len\fR for the \fIindex\fR block, and then cut
out the next \fIdata_len\fR bytes for the \fIdata\fR block.  If we have done
everything right up to this point, the following bytes would be the ASCII
formatted string \fI"XPAKSTOP"\fR.

The actual \fIdata\fR is merged into one big block; so if we want to read it,
we need the actual positions of each information in this big data block.  This
information can be obtained using the indices which are stored in the
\fIindex\fR block.

.SS Index block
The \fIindex\fR block consists of several indices:

 |<-----------------------index_len---------------------->|
 |<index1><index2><index3><index4><index5><index6><index7>|

The \fIindex\fR block holds all the information we need to find the data we
want in the \fIdata\fR block.  It consists of multiple index elements, all of
which add up to the total length \fIindex_len\fR.  It is not zero delimited
or anything else.

Each of those elements corresponds to one chunk of data in the \fIdata\fR
block: the length of that block's name (\fIname_len\fR), a \fIname_len\fR
bytes long string, the offset of that block (\fIdataN_offset\fR) and the
length of that block (\fIdataN_len\fR):

           |<name_len>|
 <name_len>|<--name-->|<dataN_offset><dataN_len>

.SS Data block
The \fIdata\fR block contains multiple chunks of data with a total length of
\fIdata_len\fR:

 |<------------------------data_len------------------------>|
 |<data1><data2><data3><data4><data5><data6><data7><data...>|

To select one data element, we need the \fIdata_offset\fR and the
\fIdata_len\fR from the \fIindex\fR.  With those, we can count
\fIdata_offset\fR bytes from the start of the \fIdata\fR block,
and then cut out the next \fIdata_len\fR bytes.  Then we got our
data block:

 |<-----dataN_offset----->|<--dataN_len->|
 |<data1data2data3data...>|<data-we-want>|
.SH EXAMPLES
Let's say we have an xpak with two chunks of data.  The first has the name
"file1" with the contents "ddDddDdd" and the second has the name "file2" with
the contents "jjJjjJjj".  There is no \fI"STOP"\fR or \fIxpak_offset\fR as
this xpak is not part of a binpkg.

Here is the hexdump output (we will break it down line by line below):
 00  58 50 41 4b 50 41 43 4b  00 00 00 20 00 00 00 10  |XPAKPACK... ....|
 10  00 00 00 04 66 69 6c 31  00 00 00 00 00 00 00 08  |....fil1........|
 20  00 00 00 04 66 69 6c 32  00 00 00 08 00 00 00 08  |....fil2........|
 30  64 64 44 64 64 44 64 64  6a 6a 4a 6a 6a 4a 6a 6a  |ddDddDddjjJjjJjj|
 40  58 50 41 4b 53 54 4f 50                           |XPAKSTOP|

The \fIindex_len\fR is 32 and the \fIdata_len\fR 16 (as there are 16 bytes:
"ddDddDdd" and "jjJjjJjj").
    |<------"XPAKPACK"----->||    32     |    16     |
 00  58 50 41 4b 50 41 43 4b  00 00 00 20 00 00 00 10

Now we have the first index element with a \fIname_len\fR of 4, followed
by the name string "fil1", followed by the data1 offset of 0 and a data1
len of 8 (since data1 has 8 bytes: "ddDddDdd").
    |     4     |<--"fil1"->||data1_off:0|data1_len:8|
 10  00 00 00 04 66 69 6c 31  00 00 00 00 00 00 00 08

Now we have the second index element with a \fIname_len\fR of 4, followed
by the name string "fil2", followed by the data2 offset of 8 and a data2
len of 8 (since data2 has 8 bytes: "jjJjjJjj").
    |     4     |<--"fil2"->||data2_off:8|data2_len:8|
 20  00 00 00 04 66 69 6c 32  00 00 00 08 00 00 00 08

    |<------"XPAKSTOP"----->|
 40  58 50 41 4b 53 54 4f 50
.SH AUTHORS
.nf
Lars Hartmann <lars@chaotika.org>
Mike Frysinger <vapier@gentoo.org>
.fi
.SH "SEE ALSO"
.BR qtbz2 (1),
.BR quickpkg (1),
.BR qxpak (1)


NEWS
portage-3.0.38.1 (2022-10-04)
--------------

Bug fixes:
* ecompress: Fix typo in ___makeopts_jobs (bug #874597).

portage-3.0.38 (2022-10-01)
--------------

Features:
* unpack now calls xz(1) with -T N where N is the number of jobs from ${MAKEOPTS}.
  This allows decompression in parallel with new enough xz (>= xz-utils 5.3.3 alpha)
  and where the source file was compressed using the parallel xz compressor
  (which becomes the default with newer xz too).

* portage: sets: Parse INCLUDES/EXCLUDES in VariableSet which allows properly
  checking for dependencies. Update @golang-rebuild accordingly which now
  includes all packages depending on dev-lang/go (bug #827974, bug #865115)

* cnf: Provide a @rust-rebuild set, similar to @golang-rebuild.

* bin: ecompress: xz, zstd: Pass -T N where N is the number of jobs from ${MAKEOPTS}
  to xz, zstd for parallel compression.

* bin: ecompress: xz: Pass -9 to xz for compression level like we do for bzip2 & gzip.

Bug fixes:
* emerge: Scheduler: Fix filenames for binpkgs in some cases - reuse existing
  filename and don't allocate a new one (bug #872392).

* dbapi: bintree: Avoid fetching 'Packages' multiple times (bug #864259).

* binpkg: Handle failure of various BINPKG_COMPRESS_* commands (bug #871570).

* data: Fix PORTAGE_USERNAME default (bug #873088).

* bin: isolated-functions.sh: return number of CPUs in ___makeopts_jobs as default
  instead of 1 to match Portage's behavior on the Python side.

* bin: ecompress: Recognize .lz, .lzo, .lz4, and .zst as compressed file
  suffixes for the purposes of the internal compressed file collision check.

* bin: ecompress: zstd: Set '-q --rm' in PORTAGE_COMPRESS_FLAGS to behave
  like other compressors.

* portage: _global_updates: Fix update summary output.

portage-3.0.37 (2022-09-23)
--------------

Bug fixes:
* _emerge: SpawnProcess: Avoid sandbox violations with PROPERTIES="interactive"
  when accessing /dev/tty* (bug #870310).

* _emerge: Binpkg: Don't create empty stray binpkgs (bug #870283).

* bin: gpkg-sign: Check gpg configuration before using (bug #869470).

* portage: gpkg: Allow missing files (needed for when ebuilds touch
  ${EROOT} in pkg_*) when creating binpkgs with quickpkg (bug #870229).

* portage: getbinpkg: Fix --getbinpkgonly command list with e.g. rsync (bug #866197).

* portage: checksum: Fix PORTAGE_CHECKSUM_FILTER. Regression from 3.0.31.

portage-3.0.36 (2022-09-09)
--------------

Bug fixes:
* The layout of gpkg (new binpkg format) packages has been fixed (bug #865905).

* Fix getbinpkg fetching (bug #864160).

portage-3.0.35 (2022-08-23)
--------------

Features:
* Portage can now set scheduling policies using the PORTAGE_SCHEDULING_POLICY
  and PORTAGE_SCHEDULING_PRIORITY variables. See make.conf.example for details.

  See also https://wiki.gentoo.org/wiki/Portage_niceness.

* The header of build.log now contains the package's slot (bug #864382).

Bug fixes:
* xml: Fix XML node parsing (resulted in duplicate USE flag descriptions
  in gentoolkit, bug 858791). This code is thought to have only one
  consumer (gentoolkit, and only as of its most recent releases), so
  impact is not severe.

* portage: dbapi: porttree: fix _set_porttrees() to operate correctly if passed
  a generator. Fixes e.g. blank emerge --metadata caches in /var/cache/edb
  which affected eix (bug #865635). Regression from 3.0.31.

* logging: don't clutter with a global logger and instead scope to
  each module. Allows further logging refactoring in future.

* estrip: fix double slash for older EAPIs (bug #862600).

* estrip: avoid spurious NEEDED warning when no ELF files installed for
  < EAPI 7 (bug #749624, bug #862606).

* estrip: apply scanelf optimisation for EAPI 7+ too (bug #749624, bug #862606).

* dispatch-conf: respect (E)ROOT

* install-qa-check.d: 60pkgconfig: drop noisy pkg-config mismatched Version
  field check (bug 857654).

* install-qa-check.d: 60pkgconfig: fix --validate check by adding
  ${ED}/usr/{lib*,share} to search path.

* install-qa-check.d: 60pkgconfig: only check defined variables for
  ${EPREFIX} compliance (bug 860825).

* install-qa-check.d: 60pkgconfig: display the detected version from
  the .pc file (improves the QA warning output), and hint to the user/developer
  about the existence of QA_PKGCONFIG_VERSION to silence the warning.

* install-qa-check.d: 60pkgconfig: use eqatag for nicer output.

* install-qa-check.d: 60pkgconfig: use ver_test where possible in version
  comparison checks to avoid warning if the Version field in the .pc file
  is equivalent to the ebuild version.

* install-qa-check.d: 80libraries: include Darwin/Mach-O objects in checks.

* install-qa-check.d: 80multilib-strict: use file/find from Prefix instead
  of absolute paths in /usr/bin.

* install-qa-checks.d: 90gcc-warnings: add arm64, aarch64 (used by Apple),
  loongarch64, riscv64, s390x to list of 64-bit arches which affects the
  severity of some warnings (like pointer truncation).

* install-qa-check.d: 90world-writable: include EPREFIX in paths.

* portage: output: update obsolete /etc/init.d/functions.sh path.

* bin: style tweaks to conform to Gentoo style (mostly variable bracing).

* doins: fix D check, add EPREFIX check

* misc-functions.sh: add Prefix fixups, like including EPREFIX in
  INSTALL_MASK for /usr/share for noman, noinfo, nodoc.

* phase-function.sh: mark ED, EROOT read-only variables

* eapi.sh: cleanup obsolete references to unofficial 5-hdepend EAPI.


lib/portage/xpak.py
# Copyright 2001-2020 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2


# The format for a tbz2/xpak:
#
#  tbz2: tar.bz2 + xpak + (xpak_offset) + "STOP"
#  xpak: "XPAKPACK" + (index_len) + (data_len) + index + data + "XPAKSTOP"
# index: (pathname_len) + pathname + (data_offset) + (data_len)
#        index entries are concatenated end-to-end.
#  data: concatenated data chunks, end-to-end.
#
# [tarball]XPAKPACKIIIIDDDD[index][data]XPAKSTOPOOOOSTOP
#
# (integer) == encodeint(integer)  ===> 4 characters (big-endian copy)
# '+' means concatenate the fields ===> All chunks are strings

__all__ = [
    "addtolist",
    "decodeint",
    "encodeint",
    "getboth",
    "getindex",
    "getindex_mem",
    "getitem",
    "listindex",
    "searchindex",
    "tbz2",
    "xpak_mem",
    "xpak",
    "xpand",
    "xsplit",
    "xsplit_mem",
]

import array
import errno

import portage
from portage import os
from portage import shutil
from portage import normalize_path
from portage import _encodings
from portage import _unicode_decode
from portage import _unicode_encode
from portage.binpkg import get_binpkg_format
from portage.exception import InvalidBinaryPackageFormat
from portage.util.file_copy import copyfile


def addtolist(mylist, curdir):
    """(list, dir) --- Takes an array(list) and appends all files from dir down
    the directory tree. Returns nothing. list is modified."""
    curdir = normalize_path(
        _unicode_decode(curdir, encoding=_encodings["fs"], errors="strict")
    )
    for parent, dirs, files in os.walk(curdir):

        parent = _unicode_decode(parent, encoding=_encodings["fs"], errors="strict")
        if parent != curdir:
            mylist.append(parent[len(curdir) + 1 :] + os.sep)

        for x in dirs:
            try:
                _unicode_decode(x, encoding=_encodings["fs"], errors="strict")
            except UnicodeDecodeError:
                dirs.remove(x)

        for x in files:
            try:
                x = _unicode_decode(x, encoding=_encodings["fs"], errors="strict")
            except UnicodeDecodeError:
                continue
            mylist.append(os.path.join(parent, x)[len(curdir) + 1 :])


def encodeint(myint):
    """Takes a 4 byte integer and converts it into a string of 4 characters.
    Returns the characters in a string."""
    a = array.array("B")
    a.append((myint >> 24) & 0xFF)
    a.append((myint >> 16) & 0xFF)
    a.append((myint >> 8) & 0xFF)
    a.append(myint & 0xFF)
    try:
        # Python >= 3.2
        return a.tobytes()
    except AttributeError:
        return a.tostring()


def decodeint(mystring):
    """Takes a 4 byte string and converts it into a 4 byte integer.
    Returns an integer."""
    myint = 0
    myint += mystring[3]
    myint += mystring[2] << 8
    myint += mystring[1] << 16
    myint += mystring[0] << 24
    return myint


def xpak(rootdir, outfile=None):
    """(rootdir, outfile) -- creates an xpak segment of the directory 'rootdir'
    and under the name 'outfile' if it is specified. Otherwise it returns the
    xpak segment."""

    mylist = []

    addtolist(mylist, rootdir)
    mylist.sort()
    mydata = {}
    for x in mylist:
        if x == "CONTENTS":
            # CONTENTS is generated during the merge process.
            continue
        x = _unicode_encode(x, encoding=_encodings["fs"], errors="strict")
        with open(os.path.join(rootdir, x), "rb") as f:
            mydata[x] = f.read()

    xpak_segment = xpak_mem(mydata)
    if outfile:
        outf = open(
            _unicode_encode(outfile, encoding=_encodings["fs"], errors="strict"), "wb"
        )
        outf.write(xpak_segment)
        outf.close()
    else:
        return xpak_segment


def xpak_mem(mydata):
    """Create an xpack segment from a map object."""

    mydata_encoded = {}
    for k, v in mydata.items():
        k = _unicode_encode(
            k, encoding=_encodings["repo.content"], errors="backslashreplace"
        )
        v = _unicode_encode(
            v, encoding=_encodings["repo.content"], errors="backslashreplace"
        )
        mydata_encoded[k] = v
    mydata = mydata_encoded
    del mydata_encoded

    indexglob = b""
    indexpos = 0
    dataglob = b""
    datapos = 0
    for x, newglob in mydata.items():
        mydatasize = len(newglob)
        indexglob = (
            indexglob
            + encodeint(len(x))
            + x
            + encodeint(datapos)
            + encodeint(mydatasize)
        )
        indexpos = indexpos + 4 + len(x) + 4 + 4
        dataglob = dataglob + newglob
        datapos = datapos + mydatasize
    return (
        b"XPAKPACK"
        + encodeint(len(indexglob))
        + encodeint(len(dataglob))
        + indexglob
        + dataglob
        + b"XPAKSTOP"
    )


def xsplit(infile):
    """(infile) -- Splits the infile into two files.
    'infile.index' contains the index segment.
    'infile.dat' contails the data segment."""
    infile = _unicode_decode(infile, encoding=_encodings["fs"], errors="strict")
    myfile = open(
        _unicode_encode(infile, encoding=_encodings["fs"], errors="strict"), "rb"
    )
    mydat = myfile.read()
    myfile.close()

    splits = xsplit_mem(mydat)
    if not splits:
        return False

    myfile = open(
        _unicode_encode(infile + ".index", encoding=_encodings["fs"], errors="strict"),
        "wb",
    )
    myfile.write(splits[0])
    myfile.close()
    myfile = open(
        _unicode_encode(infile + ".dat", encoding=_encodings["fs"], errors="strict"),
        "wb",
    )
    myfile.write(splits[1])
    myfile.close()
    return True


def xsplit_mem(mydat):
    if mydat[0:8] != b"XPAKPACK":
        return None
    if mydat[-8:] != b"XPAKSTOP":
        return None
    indexsize = decodeint(mydat[8:12])
    return (mydat[16 : indexsize + 16], mydat[indexsize + 16 : -8])


def getindex(infile):
    """(infile) -- grabs the index segment from the infile and returns it."""
    myfile = open(
        _unicode_encode(infile, encoding=_encodings["fs"], errors="strict"), "rb"
    )
    myheader = myfile.read(16)
    if myheader[0:8] != b"XPAKPACK":
        myfile.close()
        return
    indexsize = decodeint(myheader[8:12])
    myindex = myfile.read(indexsize)
    myfile.close()
    return myindex


def getboth(infile):
    """(infile) -- grabs the index and data segments from the infile.
    Returns an array [indexSegment, dataSegment]"""
    myfile = open(
        _unicode_encode(infile, encoding=_encodings["fs"], errors="strict"), "rb"
    )
    myheader = myfile.read(16)
    if myheader[0:8] != b"XPAKPACK":
        myfile.close()
        return
    indexsize = decodeint(myheader[8:12])
    datasize = decodeint(myheader[12:16])
    myindex = myfile.read(indexsize)
    mydata = myfile.read(datasize)
    myfile.close()
    return myindex, mydata


def listindex(myindex):
    """Print to the terminal the filenames listed in the indexglob passed in."""
    for x in getindex_mem(myindex):
        print(x)


def getindex_mem(myindex):
    """Returns the filenames listed in the indexglob passed in."""
    myindexlen = len(myindex)
    startpos = 0
    myret = []
    while (startpos + 8) < myindexlen:
        mytestlen = decodeint(myindex[startpos : startpos + 4])
        myret = myret + [myindex[startpos + 4 : startpos + 4 + mytestlen]]
        startpos = startpos + mytestlen + 12
    return myret


def searchindex(myindex, myitem):
    """(index, item) -- Finds the offset and length of the file 'item' in the
    datasegment via the index 'index' provided."""
    myitem = _unicode_encode(
        myitem, encoding=_encodings["repo.content"], errors="backslashreplace"
    )
    mylen = len(myitem)
    myindexlen = len(myindex)
    startpos = 0
    while (startpos + 8) < myindexlen:
        mytestlen = decodeint(myindex[startpos : startpos + 4])
        if mytestlen == mylen:
            if myitem == myindex[startpos + 4 : startpos + 4 + mytestlen]:
                # found
                datapos = decodeint(
                    myindex[startpos + 4 + mytestlen : startpos + 8 + mytestlen]
                )
                datalen = decodeint(
                    myindex[startpos + 8 + mytestlen : startpos + 12 + mytestlen]
                )
                return datapos, datalen
        startpos = startpos + mytestlen + 12


def getitem(myid, myitem):
    myindex = myid[0]
    mydata = myid[1]
    myloc = searchindex(myindex, myitem)
    if not myloc:
        return None
    return mydata[myloc[0] : myloc[0] + myloc[1]]


def xpand(myid, mydest):
    mydest = normalize_path(mydest) + os.sep
    myindex = myid[0]
    mydata = myid[1]
    myindexlen = len(myindex)
    startpos = 0
    while (startpos + 8) < myindexlen:
        namelen = decodeint(myindex[startpos : startpos + 4])
        datapos = decodeint(myindex[startpos + 4 + namelen : startpos + 8 + namelen])
        datalen = decodeint(myindex[startpos + 8 + namelen : startpos + 12 + namelen])
        myname = myindex[startpos + 4 : startpos + 4 + namelen]
        myname = _unicode_decode(
            myname, encoding=_encodings["repo.content"], errors="replace"
        )
        filename = os.path.join(mydest, myname.lstrip(os.sep))
        filename = normalize_path(filename)
        if not filename.startswith(mydest):
            # myname contains invalid ../ component(s)
            continue
        dirname = os.path.dirname(filename)
        if dirname:
            if not os.path.exists(dirname):
                os.makedirs(dirname)
        mydat = open(
            _unicode_encode(filename, encoding=_encodings["fs"], errors="strict"), "wb"
        )
        mydat.write(mydata[datapos : datapos + datalen])
        mydat.close()
        startpos = startpos + namelen + 12


class tbz2:
    def __init__(self, myfile):
        self.file = myfile
        self.filestat = None
        self.index = b""
        self.infosize = 0
        self.xpaksize = 0
        self.indexsize = None
        self.datasize = None
        self.indexpos = None
        self.datapos = None

    def decompose(self, datadir, cleanup=1):
        """Alias for unpackinfo() --- Complement to recompose() but optionally
        deletes the destination directory. Extracts the xpak from the tbz2 into
        the directory provided. Raises IOError if scan() fails.
        Returns result of upackinfo()."""
        if not self.scan():
            raise OSError
        if cleanup:
            self.cleanup(datadir)
        if not os.path.exists(datadir):
            os.makedirs(datadir)
        return self.unpackinfo(datadir)

    def compose(self, datadir, cleanup=0):
        """Alias for recompose()."""
        return self.recompose(datadir, cleanup)

    def recompose(self, datadir, cleanup=0, break_hardlinks=True):
        """Creates an xpak segment from the datadir provided, truncates the tbz2
        to the end of regular data if an xpak segment already exists, and adds
        the new segment to the file with terminating info."""
        xpdata = xpak(datadir)
        self.recompose_mem(xpdata, break_hardlinks=break_hardlinks)
        if cleanup:
            self.cleanup(datadir)

    def recompose_mem(self, xpdata, break_hardlinks=True):
        """
        Update the xpak segment.
        @param xpdata: A new xpak segment to be written, like that returned
                from the xpak_mem() function.
        @param break_hardlinks: If hardlinks exist, create a copy in order
                to break them. This makes it safe to use hardlinks to create
                cheap snapshots of the repository, which is useful for solving
                race conditions on binhosts as described here:
                https://crbug.com/185031
                Default is True.
        """
        self.scan()  # Don't care about condition... We'll rewrite the data anyway.

        if break_hardlinks and self.filestat and self.filestat.st_nlink > 1:
            tmp_fname = "%s.%d" % (self.file, portage.getpid())
            copyfile(self.file, tmp_fname)
            try:
                portage.util.apply_stat_permissions(self.file, self.filestat)
            except portage.exception.OperationNotPermitted:
                pass
            os.rename(tmp_fname, self.file)

        myfile = open(
            _unicode_encode(self.file, encoding=_encodings["fs"], errors="strict"),
            "ab+",
        )
        if not myfile:
            raise OSError
        myfile.seek(-self.xpaksize, 2)  # 0,2 or -0,2 just mean EOF.
        myfile.truncate()
        myfile.write(xpdata + encodeint(len(xpdata)) + b"STOP")
        myfile.flush()
        myfile.close()
        return 1

    def cleanup(self, datadir):
        datadir_split = os.path.split(datadir)
        if len(datadir_split) >= 2 and len(datadir_split[1]) > 0:
            # This is potentially dangerous,
            # thus the above sanity check.
            try:
                shutil.rmtree(datadir)
            except OSError as oe:
                if oe.errno == errno.ENOENT:
                    pass
                else:
                    raise oe

    def scan(self):
        """Scans the tbz2 to locate the xpak segment and setup internal values.
        This function is called by relevant functions already."""
        a = None
        try:
            mystat = os.stat(self.file)
            if self.filestat:
                changed = 0
                if (
                    mystat.st_size != self.filestat.st_size
                    or mystat.st_mtime != self.filestat.st_mtime
                    or mystat.st_ctime != self.filestat.st_ctime
                ):
                    changed = True
                if not changed:
                    return 1
            self.filestat = mystat
            a = open(
                _unicode_encode(self.file, encoding=_encodings["fs"], errors="strict"),
                "rb",
            )
            a.seek(-16, 2)
            trailer = a.read()
            self.infosize = 0
            self.xpaksize = 0
            if trailer[-4:] != b"STOP":
                try:
                    get_binpkg_format(self.file, check_file=True)
                except InvalidBinaryPackageFormat:
                    pass
                return 0
            if trailer[0:8] != b"XPAKSTOP":
                try:
                    get_binpkg_format(self.file, check_file=True)
                except InvalidBinaryPackageFormat:
                    pass
                return 0
            self.infosize = decodeint(trailer[8:12])
            self.xpaksize = self.infosize + 8
            a.seek(-(self.xpaksize), 2)
            header = a.read(16)
            if header[0:8] != b"XPAKPACK":
                try:
                    get_binpkg_format(self.file, check_file=True)
                except InvalidBinaryPackageFormat:
                    pass
                return 0
            self.indexsize = decodeint(header[8:12])
            self.datasize = decodeint(header[12:16])
            self.indexpos = a.tell()
            self.index = a.read(self.indexsize)
            self.datapos = a.tell()
            return 2
        except SystemExit:
            raise
        except:
            try:
                get_binpkg_format(self.file, check_file=True)
            except InvalidBinaryPackageFormat:
                pass
            return 0
        finally:
            if a is not None:
                a.close()

    def filelist(self):
        """Return an array of each file listed in the index."""
        if not self.scan():
            return None
        return getindex_mem(self.index)

    def getfile(self, myfile, mydefault=None):
        """Finds 'myfile' in the data segment and returns it."""
        if not self.scan():
            return None
        myresult = searchindex(self.index, myfile)
        if not myresult:
            return mydefault
        a = open(
            _unicode_encode(self.file, encoding=_encodings["fs"], errors="strict"), "rb"
        )
        a.seek(self.datapos + myresult[0], 0)
        myreturn = a.read(myresult[1])
        a.close()
        return myreturn

    def getelements(self, myfile):
        """A split/array representation of tbz2.getfile()"""
        mydat = self.getfile(myfile)
        if not mydat:
            return []
        return mydat.split()

    def unpackinfo(self, mydest):
        """Unpacks all the files from the dataSegment into 'mydest'."""
        if not self.scan():
            return 0
        mydest = normalize_path(mydest) + os.sep
        a = open(
            _unicode_encode(self.file, encoding=_encodings["fs"], errors="strict"), "rb"
        )
        if not os.path.exists(mydest):
            os.makedirs(mydest)
        startpos = 0
        while (startpos + 8) < self.indexsize:
            namelen = decodeint(self.index[startpos : startpos + 4])
            datapos = decodeint(
                self.index[startpos + 4 + namelen : startpos + 8 + namelen]
            )
            datalen = decodeint(
                self.index[startpos + 8 + namelen : startpos + 12 + namelen]
            )
            myname = self.index[startpos + 4 : startpos + 4 + namelen]
            myname = _unicode_decode(
                myname, encoding=_encodings["repo.content"], errors="replace"
            )
            filename = os.path.join(mydest, myname.lstrip(os.sep))
            filename = normalize_path(filename)
            if not filename.startswith(mydest):
                # myname contains invalid ../ component(s)
                continue
            dirname = os.path.dirname(filename)
            if dirname:
                if not os.path.exists(dirname):
                    os.makedirs(dirname)
            mydat = open(
                _unicode_encode(filename, encoding=_encodings["fs"], errors="strict"),
                "wb",
            )
            a.seek(self.datapos + datapos)
            mydat.write(a.read(datalen))
            mydat.close()
            startpos = startpos + namelen + 12
        a.close()
        return 1

    def get_data(self):
        """Returns all the files from the dataSegment as a map object."""
        if not self.scan():
            return {}
        a = open(
            _unicode_encode(self.file, encoding=_encodings["fs"], errors="strict"), "rb"
        )
        mydata = {}
        startpos = 0
        while (startpos + 8) < self.indexsize:
            namelen = decodeint(self.index[startpos : startpos + 4])
            datapos = decodeint(
                self.index[startpos + 4 + namelen : startpos + 8 + namelen]
            )
            datalen = decodeint(
                self.index[startpos + 8 + namelen : startpos + 12 + namelen]
            )
            myname = self.index[startpos + 4 : startpos + 4 + namelen]
            a.seek(self.datapos + datapos)
            mydata[myname] = a.read(datalen)
            startpos = startpos + namelen + 12
        a.close()
        return mydata

    def getboth(self):
        """Returns an array [indexSegment, dataSegment]"""
        if not self.scan():
            return None

        a = open(
            _unicode_encode(self.file, encoding=_encodings["fs"], errors="strict"), "rb"
        )
        a.seek(self.datapos)
        mydata = a.read(self.datasize)
        a.close()

        return self.index, mydata
